File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
from collections import deque
n, m = map(int, input().split())
graph = []
for _ in range(n):
    graph.append(list(map(int, input())))

visited = [[0 for _ in range(m)] for _ in range(n)]

dx=[-1,0,1,0]
dy=[0,1,0,-1]


def bfs(x,y):
  queue = deque()
  queue.append((x,y))
  
  while queue:
    x,y=queue.popleft()
    if(visited[x][y]==0):
      visited[x][y]=1
      for i in range(4):
        nx = x+dx[i]
        ny = y+dy[i]
        if(nx>=n or ny>=m or nx<0 or ny<0):
          continue
        if(visited[nx][ny]==0 and graph[nx][ny]==1):
          queue.append((nx,ny))
          graph[nx][ny]=graph[x][y]+1
          


bfs(0,0)
print(graph[n-1][m-1])

File: 백준/Silver/2178. 미로 탐색/미로 탐색.cc
#include <iostream>
#include <queue>

int main()
{
	int N, M;
	int graph[100][100] = { 0, };
	int visited[100][100] = { 0, };
	std::queue<std::pair<int, int>> Q;

	int dx[4] = { 0, 0, -1, 1 };
	int dy[4] = { 1,-1,0,0 };


	std::cin >> N >> M;

	for (int i = 0; i < N; ++i)
	{
		std::string row;
		std::cin >> row;

		for (int j = 0; j < M; ++j)
		{
			if (row[j] == '1')
			{
				graph[i][j] = 1;
			}
		}
	}

	Q.push(std::make_pair(0, 0));
	visited[0][0] = 1;
	while (!Q.empty())
	{
		int x = Q.front().first;
		int y = Q.front().second;
		Q.pop();
		for (int dir = 0; dir < 4; ++dir)
		{
			int nx = x + dx[dir];
			int ny = y + dy[dir];
			if (0 == graph[nx][ny] || 0 < visited[nx][ny] || nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
			Q.push(std::make_pair(nx, ny));
			visited[nx][ny] = visited[x][y] + 1;
		}
	}


	std::cout << visited[N - 1][M - 1];
	return 0;
}